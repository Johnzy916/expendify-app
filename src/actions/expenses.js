import uuid from 'uuid';
import database from '../firebase/firebase.js';

// ADD_EXPENSES ACTION
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// WRITE EXPENSE TO DATABASE / DISPATCH ACTION
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense)
      .then(ref => {
        dispatch(addExpense({
          id: ref.key, 
          ...expense 
        }));
      }).catch(error => {
        console.log('Couldn\'t add expense > ', error);
      });
  };
};

// REMOVE_EXPENSE ACTION
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE ACTION
export const editExpense = (id, edits) => ({
  type: 'EDIT_EXPENSE',
  id,
  edits
});

// SET_EXPENSE ACTION
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// FETCH EXPENSES FROM DATABASE / DISPATCH ACTION
 export const startSetExpenses = () => {
   return (dispatch) => {
     return database.ref('expenses').once('value')
    .then(snapshot => {
      const expenses = [];
      snapshot.forEach(expense => {
        expenses.push({
          id: expense.key,
          ...expense.val()
        });
      });
      dispatch(setExpenses(expenses));
    }).catch(error => console.log('couldn\'t get expenses > : ', error));
   };
 };

