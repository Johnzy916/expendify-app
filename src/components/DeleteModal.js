import React from 'react';
import ReactModal from 'react-modal';
ReactModal.setAppElement('#app');

const DeleteModal = ({ showModal, handleConfirmDelete, handleCancelDelete }) => (
    <ReactModal
        isOpen={showModal}
        contentLabel="Delete expense"
        onRequestClose={handleCancelDelete}
        closeTimeoutMS={200}
        className="react-modal"
    >
        <h3 className="react-modal__body">
            Are you sure you want to delete this expense?
        </h3>
        <button
            className="btn btn--primary btn--shine react-modal__btn"
            onClick={handleCancelDelete}
        >
            Keep expense
        </button>
        <button 
            className="btn btn--tertiary btn--shine react-modal__btn"
            onClick={handleConfirmDelete}
        >
            Yes, delete
        </button>
    </ReactModal>
);

export default DeleteModal;
