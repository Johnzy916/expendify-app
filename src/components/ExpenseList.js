import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
  <div>
    {
      expenses.length === 0 ? (
        <p>No Expenses</p>
      ) : (
        expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )
    }
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);