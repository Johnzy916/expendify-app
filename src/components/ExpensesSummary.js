import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/selectExpensesTotal';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => (
  <p>
  Viewing {expenseCount}&nbsp;
  {expenseCount === 1 ? "expense" : "expenses"}&nbsp;
  totalling {numeral(expensesTotal / 100).format('$0,0.00')}
  </p>
);

const mapStateToProps = state => {
  const expenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expensesTotal: selectExpensesTotal(expenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
