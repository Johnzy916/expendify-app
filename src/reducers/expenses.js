const expensesReducer = (state, action) => {
  if (typeof state === 'undefined') { 
    return [] 
  }
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.edits
          };
        } else { return expense };
      });
    default:
      return state;
  }
}

export default expensesReducer;