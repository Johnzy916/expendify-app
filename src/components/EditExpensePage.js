import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ReactModal from './DeleteModal';

export class EditExpensePage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showModal: false
    }
  }
  
  onSubmit = editedExpense => {
    this.props.startEditExpense(this.props.expense.id, editedExpense);
    this.props.history.push('/');
  }
  
  hideModal = () => {
    this.setState(() => ({
      showModal: false
    }));
  }
  
  onRemove = () => {
    this.setState(() => ({
      showModal: true
    }))
  }
  
  handleConfirmDelete = () => {
    this.hideModal();
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  
  handleCancelDelete = () => {
    this.hideModal();
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">
              Edit Expense
            </h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
            submitType="Save"
          />
          <button
            className="btn btn--tertiary btn--shine"
            onClick={this.onRemove}
          >
            Remove
          </button>
        </div>
        <ReactModal 
          showModal={this.state.showModal}
          handleConfirmDelete={this.handleConfirmDelete}
          handleCancelDelete={this.handleCancelDelete}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, editedExpense) => dispatch(startEditExpense(id, editedExpense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);