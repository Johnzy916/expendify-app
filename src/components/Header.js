import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom'
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout, userName }) => (
  <header className="header">
      <div className="content-container">
        <nav className="header__content">
          <Link
            className="header__title"
            to="/dashboard">
              <h1>Expendify</h1>
          </Link>
          <div>
            <span className="header__username">
              {userName}
            </span>
            <button
              className="btn btn--primary btn--link"
              onClick={startLogout}>
                Logout
            </button>
          </div>
        </nav>
      </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = state => ({
  userName: state.auth.userName
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);