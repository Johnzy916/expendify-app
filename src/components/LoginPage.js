import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expendify</h1>
            <p>Track, search, and expect what you're spending.</p>
            <button 
                className="btn btn--google btn--shine"
                onClick={startLogin}>
                    Login with Google
            </button>
        </div>
    </div>
);

const mapDispatchToProps = dispatch => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);