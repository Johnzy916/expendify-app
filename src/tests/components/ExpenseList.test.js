import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
import '../setupTests.js'; // needed for codesandbox.io
import toJSON from 'enzyme-to-json'; // needed for codesandbox.io


test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});