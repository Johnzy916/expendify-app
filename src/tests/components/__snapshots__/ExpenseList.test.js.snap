// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={123}
    createdAt={345600000}
    description="Pickles"
    id="1"
    key="1"
    note="pickles"
  />
  <ExpenseListItem
    amount={234}
    createdAt={-345600000}
    description="Lettuce"
    id="2"
    key="2"
    note="lettuce"
  />
  <ExpenseListItem
    amount={456}
    createdAt={0}
    description="Tomato"
    id="3"
    key="3"
    note="tomato"
  />
</div>
`;
